account_cash_discount_base/models/account_invoice.py

```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+# Copyright 2014 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from datetime import datetime, timedelta
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+READONLY_STATES = {
+    'draft': [('readonly', False)],
+}
+
+
+class AccountInvoice(models.Model):
+
+    _inherit = 'account.invoice'
+
+    discount_percent = fields.Float(
+        string="Discount (%)",
+        readonly=True,
+        states=READONLY_STATES,
+    )
+    amount_total_with_discount = fields.Monetary(
+        string="Total (with discount)",
+        compute='_compute_amount_total_with_discount',
+    )
+    discount_amount = fields.Monetary(
+        compute='_compute_discount_amount',
+    )
+    discount_delay = fields.Integer(
+        string="Discount Delay (days)",
+        readonly=True,
+        states=READONLY_STATES
+    )
+    discount_due_date = fields.Date(
+        readonly=True,
+        states=READONLY_STATES
+    )
+    discount_due_date_readonly = fields.Date(
+        string="Discount Due Date",
+        compute='_compute_discount_due_date',
+    )
+
+    @api.multi
+    @api.depends(
+        'amount_total',
+        'amount_untaxed',
+        'discount_percent',
+        'company_id.cash_discount_base_amount_type',
+    )
+    def _compute_discount_amount(self):
+        for rec in self:
+            discount_amount = 0.0
+            if rec.discount_percent != 0.0:
+                base_amount_type = \
+                    rec.company_id.cash_discount_base_amount_type
+                base_amount = (
+                    base_amount_type == 'untaxed' and
+                    rec.amount_untaxed or rec.amount_total)
+                discount_amount = (
+                    base_amount * (0.0 + rec.discount_percent / 100))
+            rec.discount_amount = discount_amount
+
+    @api.multi
+    @api.depends(
+        'amount_total',
+        'discount_amount',
+    )
+    def _compute_amount_total_with_discount(self):
+        for rec in self:
+            rec.amount_total_with_discount = \
+                rec.amount_total - rec.discount_amount
+
+    @api.multi
+    @api.depends(
+        'discount_due_date',
+    )
+    def _compute_discount_due_date(self):
+        for rec in self:
+            rec.discount_due_date_readonly = rec.discount_due_date
+
+    @api.multi
+    @api.onchange(
+        'discount_delay',
+    )
+    def _onchange_discount_delay(self):
+        date_today = datetime.today()
+        for rec in self:
+            if rec.discount_amount == 0 or rec.discount_delay == 0:
+                continue
```

Comment:
I should check the ```discount_percent``` instead of the ```discount_amount```. This test was to prevent to compute a discount due date if it was not needed.

But I'm wondering if we should add a boolean field which would tells if we pay the invoice using the discount or not. If this boolean equals to True, we could maybe force the user to fill the discount_percent and discount_delay.